File: MRouter-APT/src/main/java/com/whx/router/compiler/utils/Logger.java
package com.whx.router.compiler.utils;

import static com.whx.router.compiler.utils.Constants.PREFIX_OF_LOGGER;

import org.apache.commons.lang3.StringUtils;

import javax.annotation.processing.Messager;
import javax.tools.Diagnostic;

public class Logger {
    private final Messager msg;

    public Logger(Messager messager) {
        msg = messager;
    }

    public void info(CharSequence info) {
        if (StringUtils.isNotEmpty(info)) {
            msg.printMessage(Diagnostic.Kind.NOTE, Constants.PREFIX_OF_LOGGER + info);
        }
    }

    public void error(CharSequence error) {
        if (StringUtils.isNotEmpty(error)) {
            msg.printMessage(Diagnostic.Kind.ERROR, Constants.PREFIX_OF_LOGGER + "An exception is encountered, [" + error + "]");
        }
    }

    public void error(Throwable error) {
        if (null != error) {
            msg.printMessage(Diagnostic.Kind.ERROR, Constants.PREFIX_OF_LOGGER + "An exception is encountered, [" + error.getMessage() + "]" + "\n" + formatStackTrace(error.getStackTrace()));
        }
    }

    public void warning(CharSequence warning) {
        if (StringUtils.isNotEmpty(warning)) {
            msg.printMessage(Diagnostic.Kind.WARNING, Constants.PREFIX_OF_LOGGER + warning);
        }
    }

    private String formatStackTrace(StackTraceElement[] stackTrace) {
        StringBuilder sb = new StringBuilder();
        for (StackTraceElement element : stackTrace) {
            sb.append("    at ").append(element.toString());
            sb.append("\n");
        }
        return sb.toString();
    }
}

File: MRouter-Plugin/src/main/kotlin/com/whx/router/plugin/utils/Logger.kt
package com.whx.router.plugin.utils

import com.whx.router.plugin.utils.Constants.PROJECT

object Logger {

    fun i(tag: String, info: String) {
        println("[info] ${PROJECT}::Gradle-Plugin >>> TAG:$tag $info")
    }

    fun w(tag: String, info: String) {
        println("[warning] ${PROJECT}::Gradle-Plugin >>> TAG:$tag $info")
    }

    fun e(tag: String, info: String) {
        error("[error] ${PROJECT}::Gradle-Plugin >>> TAG:$tag $info")
    }

}